### ISSUE TITLE
[Feature] Create a "Draft" pull request

### ISSUE THREAD

**Issue by stephenyeargin** [github/hub#2038]
What better time to request a feature in `hub` than moments after a feature is released in the GitHub UI! 😁 I love not having to go to the UI for common tasks, and I can see this new feature being included in my workflow for discussing a PR prior to wanting to merge it.

Reference: https://github.blog/2019-02-14-introducing-draft-pull-requests/

A cursory look at the API docs doesn't seem to have this available yet (I do see a "locked" flag), so this issue may hang out for a while.

**Comment by dwijnand** [comment]
Looks like the draft attribute on pull requests is now documented as available under `Accept: application/vnd.github.shadow-cat-preview`:

https://developer.github.com/v3/pulls/#create-a-pull-request

**Comment by omgjlk** [comment]
I'm taking a shot at implementing this.

**Comment by omgjlk** [comment]
Hrm, that's API only for SEEING if a PR is in draft mode, it's not the same as SETTING a PR to draft mode. Seeking more clarification.

**Comment by omgjlk** [comment]
Confirmed, for REST API (which is what `hub` makes use of) there is only support for seeing draft status in responses, not for setting draft status.

**Comment by mislav** [comment]
@omgjlk Thank you for looking into this! I guess we'll have to wait for the API to be ready. Until then we could at least expose the `draft` field when listing PRs.

**Comment by jrochkind** [comment]
I would love the ability to do this too. I wonder if there's a way to make Github aware of our feature request?  I never use web UI PR anymore, I always use `hub`, and it's annoying not to have access to the new "draft" feature which is very useful.

**Comment by omgjlk** [comment]
GitHub is aware 😀

**Comment by bbugh** [comment]
Looking forward to this being built in!

In the meantime for anybody missing this feature, here's a bash function that can do something similar:

```bash
function gpr() {
    local branch=$(git rev-parse --abbrev-ref HEAD)
    local base=${1:-develop}
    local repo=$(git remote get-url origin)
    open $repo/compare/$base...$branch
}
```

Use it like:

  - `$ gpr` - opens a new pull request on github.com for the current branch and the base "develop" (you should change this to whatever your default branch is)
  - `$ gpr base` - opens a new PR for the current branch to the `base` branch.

**Comment by jrochkind** [comment]
@bbugh I'm confused, have you found a way to have a command line that opens a PR in Github as a  "Draft" PR?

That's what we're missing. The `hub` tool already lets you open a PR from the command line, no problem, just not as a "Draft", using the new "Draft"... functionality/mode/not-sure-what-they-call-this that GH has added.

**Comment by bbugh** [comment]
WHOOPS. Ignore me!

**Comment by dwijnand** [comment]
No, @bbugh, having a script that opens GitHub's compare/x...y page (in which you can finish the PR process and select draft) is useful!

**Comment by jrochkind** [comment]
Ah, I get it. 

Would it make sense to add this to `hub`? It might be useful on it's own, and would also serve as a workaround to the missing "draft" feature in hub. Still gotta go to the web, but it's fewer clicks if it had this feature. 

@bbugh I'd encourage you to send a PR!  (I am not a maintainer or committer, just another user frustrated that `hub pull-request` doesn't have the "draft" feature!)

**Comment by ndench** [comment]
You can compare a branch with `hub compare` and create the @dwijnand craft pull request yourself.

**Comment by dwijnand** [comment]
```bash
git config --global alias.pr-draft '!hub compare $(git rev-parse --symbolic-full-name --abbrev-ref HEAD)'
```

**Comment by mislav** [comment]
@dwijnand Thank you for sharing! Shouldn't `hub compare` without any arguments already open the compare page for the current branch?

**Comment by dwijnand** [comment]
Perhaps it should (but currently doesn't).

**Comment by kheyse-oqton** [comment]
I'd like to link this feature request to this PR to allow using github's own interface to create a PR: https://github.com/github/hub/pull/1005
If this would be available, we would not be so dependent on the hub pull-request command to expose new features of github, like the draft PRs.

**Comment by omgjlk** [comment]
> Perhaps it should (but currently doesn't).

Does it not when you are working from a fork? It seems to work when the branch I'm working on exists in the same repo, rather than a fork. I haven't tried on a fork lately.

**Comment by dwijnand** [comment]
> Does it not when you are working from a fork? It seems to work when the branch I'm working on exists in the same repo, rather than a fork. I haven't tried on a fork lately.

How does it work for you?  For me without arguments it just fast-fails:

```bash
$ git compare
Usage: hub compare [-uc] [<USER>] [[<START>...]<END>]
       hub compare [-uc] [-b <BASE>]
```

**Comment by kheyse-oqton** [comment]
I get this output when I haven't pushed my branch yet. Some meaningful feedback would be useful in this case.

**Comment by omgjlk** [comment]
I've only tried after pushing my branch, so maybe that's the key. When I run the command it just opens my browser window to the appropriate view.

**Comment by mislav** [comment]
@dwijnand This perhaps warrants its own discussion in a spearate thread (also please see https://github.com/github/hub/issues/1808), but if `hub compare` doesn't open the compare page for the current branch, then please:

1. Check if you pushed the current branch anywhere
2. Open an issue detailing your git remote, `git config push.default`, and branch configuration

Thank you!

**Comment by ThiefMaster** [comment]
Could this also be done for converting an issue to a PR? Especially since the UI doesn't let you convert a normal PR to draft, being able to convert an issue to a draft PR would be very useful!

**Comment by auscompgeek** [comment]
> Could this also be done for converting an issue to a PR?

My understanding was that issue-to-PR conversion was completely removed a while ago. Does it still work?

**Comment by ThiefMaster** [comment]
yes, and the API also isn't marked as deprecated anymore

**Comment by dwijnand** [comment]
TIL! https://github.com/github/hub/issues/1927

**Comment by jasonkarns** [comment]
The API now supports creating Draft PRs!!! https://github.blog/changelog/2019-03-26-create-a-draft-pull-request-from-the-rest-api/

**Comment by rsinnet** [comment]
> The API now supports creating Draft PRs!!! https://github.blog/changelog/2019-03-26-create-a-draft-pull-request-from-the-rest-api/

Is anyone actively working on this?

**Comment by thehunt33r** [comment]
The feature is still in preview, therefore implementing this requires adding the `Accept` header to `application/vnd.github.shadow-cat-preview`.

However I've got a working version on my [fork](https://github.com/thehunt33r/hub/tree/Implement-Draft). I won't open a PR until this feature is out of preview though, as I don't think `hub` should rely on experimental features.

**Comment by mislav** [comment]
This is now [available in 2.11.0](https://github.com/github/hub/releases/tag/v2.11.0); thank you @thehunt33r for the implementation.

Hub [already uses some APIs that are under "preview"](https://github.com/github/hub/commit/ecaafce40787b684c4754c1a17c2070dd749d1e8); we don't shy away from that. Historically, we never had a problem with using experimental API features.

When listing PRs, you can use the `%pS` custom format parameter to list the state of the PR, and one of these states can be "draft". Example:

```sh
hub pr list -L20 --format='%i %pC%<(7)%pS%Creset %t%n' --state all
```

**Comment by gbrlsnchs** [comment]
It seems both `--draft` and `-d` are missing from ZSH completion for Pull Requests. :disappointed:
